### PR TITLE
[#137159] add training requested filter to bulk emailer

### PR DESCRIPTION
**Pull request by meara**
https://pm.tablexi.com/issues/137159

### CODE REVIEW

File: vendor/engines/bulk_email/app/assets/javascripts/bulk_email/bulk_email.js.coffee [hunk 1]
```diff
@@ -9,6 +9,23 @@ class window.BulkEmailSearchForm
 
   selectedUserTypes: -> @$userTypeCheckboxes().filter(':checked').map -> @.value
 
+  dateIrrelevantSelectedOnly: ->
+    user_types = @selectedUserTypes().toArray()
+    (user_types.includes('authorized_users') || user_types.includes('training_requested')) &&
```
Comment (jhanggi) [inline vendor/engines/bulk_email/app/assets/javascripts/bulk_email/bulk_email.js.coffee]: This is kind of a confusing and very verbose check.

What if we add a data attributes to these fields  marking them as date relevant/irrelevant and then look to make sure our logic passes? I'd prefer if we can avoid that mapping in the JS if we can.

File: vendor/engines/bulk_email/app/services/bulk_email/recipient_searcher.rb [hunk 2]
```diff
@@ -62,6 +62,14 @@ def search_authorized_users
         .reorder(*self.class::DEFAULT_SORT)
     end
 
+    def search_training_requested
+      users
+        .joins(:training_requests)
+        .uniq
+        .where(training_requests: { product_id: product_ids_from_params })
```
Comment (jhanggi) [inline vendor/engines/bulk_email/app/services/bulk_email/recipient_searcher.rb]: I think we handle it, but can we make sure this does not include users who issued a training request, but then were later approved? We actually destroy training requests on the "destroy" action, so this shouldn't include users who's request was denied.

File: vendor/engines/bulk_email/app/assets/javascripts/bulk_email/bulk_email.js.coffee [hunk 3]
```diff
@@ -3,33 +3,45 @@ class window.BulkEmailSearchForm
     @_initUserTypeChangeHandler()
     @_initDateRangeSelectionHandlers()
 
-  authorizedUsersSelectedOnly: ->
+  hideNonRestrictedProducts: ->
     user_types = @selectedUserTypes()
-    user_types.length == 1 && user_types[0] == 'authorized_users'
+    required_selected = user_types.filter(['authorized_users', 'training_requested'])
+    # if either authorized_users or training_requested is selected (and nothing else), we should display only restricted products
```
Comment (jhanggi) [inline vendor/engines/bulk_email/app/assets/javascripts/bulk_email/bulk_email.js.coffee]: Maybe split the comment? I'm seeing a lint violation.

**Comment by meara** [comment]
@jhanggi what kind of tests were you thinking? I thought we had decided unit testing the javascript wasn't worth the time required to set up?

**Comment by jhanggi** [comment]
Let's not test the JS piece around disabling/enable other fields, but let's at least test that if you choose the new checkbox, it returns the people we expect.

**Pull request merged**